src/posix/platform/multicast_routing.cpp

```diff
@@ -0,0 +1,563 @@
+}
+
+otError MulticastRoutingManager::AddMulticastForwardingCache(const Ip6::Address &aSrcAddr,
+                                                             const Ip6::Address &aGroupAddr,
+                                                             MifIndex            aIif)
+{
+    otError        error = OT_ERROR_NONE;
+    struct mf6cctl mf6cctl;
+    MifIndex       forwardMif = kMifIndexNone;
+
+    VerifyOrExit(aIif == kMifIndexThread || aIif == kMifIndexBackbone, error = OT_ERROR_INVALID_ARGS);
+
+    ExpireMulticastForwardingCache();
+
+    if (aIif == kMifIndexBackbone)
+    {
+        // Forward multicast traffic form Backbone to Thread if the group address is subscribed by MLR.
+        if (mListenerSet.find(aGroupAddr) != mListenerSet.end())
+        {
+            forwardMif = kMifIndexThread;
+        }
+    }
+    else
+    {
+        // Forward multicast traffic from Thread to Backbone if multicast scope > kRealmLocalScope
+        if (aGroupAddr.GetScope() > Ip6::Address::kRealmLocalScope)
+        {
+            forwardMif = kMifIndexBackbone;
+        }
+    }
+
+    memset(&mf6cctl, 0, sizeof(mf6cctl));
+
+    memcpy(mf6cctl.mf6cc_origin.sin6_addr.s6_addr, aSrcAddr.GetBytes(), sizeof(mf6cctl.mf6cc_origin.sin6_addr.s6_addr));
+    memcpy(mf6cctl.mf6cc_mcastgrp.sin6_addr.s6_addr, aGroupAddr.GetBytes(),
+           sizeof(mf6cctl.mf6cc_mcastgrp.sin6_addr.s6_addr));
+    mf6cctl.mf6cc_parent = aIif;
+
+    if (forwardMif != kMifIndexNone)
+    {
+        IF_SET(forwardMif, &mf6cctl.mf6cc_ifset);
+    }
+
+    VerifyOrExit(0 == setsockopt(mMulticastRouterSock, IPPROTO_IPV6, MRT6_ADD_MFC, &mf6cctl, sizeof(mf6cctl)),
+                 error = OT_ERROR_FAILED);
+
+    mMulticastForwardingCache[MulticastRoute(aSrcAddr, aGroupAddr)] = MulticastRouteInfo(aIif, forwardMif);
+
+exit:
+    otLogResultPlat(error, "MulticastRoutingManager: %s: add dynamic route: %s %s => %s %s", __FUNCTION__,
+                    MifIndexToString(aIif), aSrcAddr.ToString().AsCString(), aGroupAddr.ToString().AsCString(),
+                    MifIndexToString(forwardMif));
+
+    return error;
+}
+
+void MulticastRoutingManager::UnblockInboundMulticastForwardingCache(const Ip6::Address &aGroupAddr)
+{
+    struct mf6cctl mf6cctl;
+
+    memset(&mf6cctl, 0, sizeof(mf6cctl));
+    memcpy(mf6cctl.mf6cc_mcastgrp.sin6_addr.s6_addr, aGroupAddr.GetBytes(),
+           sizeof(mf6cctl.mf6cc_mcastgrp.sin6_addr.s6_addr));
+    mf6cctl.mf6cc_parent = kMifIndexBackbone;
+    IF_SET(kMifIndexThread, &mf6cctl.mf6cc_ifset);
+
+    for (const std::pair<const MulticastRoute, MulticastRouteInfo> &entry : mMulticastForwardingCache)
+    {
+        otError                   error;
+        const MulticastRoute &    route     = entry.first;
+        const MulticastRouteInfo &routeInfo = entry.second;
+
+        if (routeInfo.mIif != kMifIndexBackbone || routeInfo.mOif == kMifIndexThread || route.mGroupAddr != aGroupAddr)
+        {
+            continue;
+        }
```

Comment:
what's the purpose of this `for` loop?
When adding new listener, it seems to me not possible that such entry exist already.